### PR TITLE
Convert editor to TypeScript

### PR DESCRIPTION
**Pull request by t3chguy**

### CODE REVIEW

File: src/editor/serialize.ts [hunk 1]
```diff
@@ -31,7 +32,7 @@ export function mdSerialize(model: EditorModel) {
                 return html + part.text;
             case "room-pill":
             case "user-pill":
-                return html + `[${part.text.replace(/[[\\\]]/g, c => "\\" + c)}](${makeGenericPermalink(part.resourceId)})`;
+                return html + `[${part.text.replace(/[[\\\]]/g, c => "\\" + c)}](${makeGenericPermalink((part as PillPart).resourceId)})`;
```
Comment (JorikSchellekens) [inline src/editor/serialize.ts:35]: Is this the only place we expect to switch over part types? Is it worth expressing them as a discriminated union?

File: src/editor/offset.ts [hunk 2]
```diff
@@ -14,17 +14,17 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+import EditorModel from "./model";
+
 export default class DocumentOffset {
-    constructor(offset, atNodeEnd) {
-        this.offset = offset;
-        this.atNodeEnd = atNodeEnd;
+    constructor(public offset: number, public readonly atNodeEnd: boolean) {
```
Comment (JorikSchellekens) [inline src/editor/offset.ts:20]: Are we cool with this style?

File: src/editor/autocomplete.ts [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+    constructor(
+        private updateCallback: UpdateCallback,
+        private getAutocompleterComponent: GetAutocompleterComponent,
+        private updateQuery: UpdateQuery,
+        private partCreator: PartCreator | CommandPartCreator,
+    ) {
+    }
+
+    onEscape(e: KeyboardEvent) {
```
Comment (JorikSchellekens) [inline src/editor/autocomplete.ts:46]: Can we public these

File: src/components/views/rooms/BasicMessageComposer.tsx [hunk 4]
```diff
@@ -71,45 +77,75 @@ function selectionEquals(a: Selection, b: Selection): boolean {
...
+
+        this.emoticonSettingHandle = SettingsStore.watchSetting('MessageComposerInput.autoReplaceEmoji', null,
+            this.configureEmoticonAutoReplace);
+        this.configureEmoticonAutoReplace();
+        this.shouldShowPillAvatarSettingHandle = SettingsStore.watchSetting("Pill.shouldShowPillAvatar", null,
+            this.configureShouldShowPillAvatar);
     }
 
-    componentDidUpdate(prevProps) {
+    componentDidUpdate(prevProps: IProps) {
```
Comment (JorikSchellekens) [inline src/components/views/rooms/BasicMessageComposer.tsx:137]: public

File: src/editor/offset.ts (hunk 2 above)
Comment (t3chguy) [reply src/editor/offset.ts:20]: we already use it in a few places, it is so much less chatty thankfully

File: src/editor/serialize.ts (hunk 1 above)
Comment (t3chguy) [reply src/editor/serialize.ts:35]: That is a grand idea!

Comment (t3chguy) [reply src/editor/serialize.ts:35]: okay that was a PAIN because class extensions don't play well with it :D

Comment (JorikSchellekens) [reply src/editor/serialize.ts]: Sorry for putting you through it

Comment (t3chguy) [reply src/editor/serialize.ts]: worth it in the long term

**Pull request merged**